### ISSUE TITLE
UCell support instead of AddModuleScore?

### ISSUE THREAD

**Issue by cnk113** [enblacar/SCpubr#9]
Hello,

This is a great package! I just was wondering if it was possible to support UCell scoring functionality then AddModuleScore?
It's a bit more scalable and has different scoring heuristics so it would be great to have that as an option.

Thanks,
Chang

**Comment by enblacar** [comment]
Hi @cnk113,

Thanks for the interest and the feedback! Happy to hear that you like the package.

This is indeed a very interesting addition. I guess you mean using it as scoring method in `SCpubr::do_EnrichmentHeatmap()` no?

I will have a look at it and put in the To-Do list!

Best,
Enrique

**Comment by enblacar** [comment]
Hi @cnk113,

I have implemented the method in the last commit. Could you please check it out and tell me if it works for you as intended?

```r
devtools::install_github("enblacar/SCpubr", ref = "d6e7617a9c4e7817da8bf9bfde73eb7634991625")
```

For this, use the following code:

```r
SCpubr::do_EnrichmentHeatmap(sample = sample, 
                             input_gene_list = genes, 
                             ncores = 1, 
                             flavor = "UCell", 
                             storeRanks = TRUE)
```

With it, I got the following plot:
![e419b6cc-b883-4c60-9e4b-f4cbdcbce610](https://user-images.githubusercontent.com/42321264/199717515-cc256a5c-9dd0-45c4-8e27-e438c502866c.png)

While with Seurat's `AddModuleScore`:
![cca4415d-5957-4d6a-946a-4ab55fcd9396](https://user-images.githubusercontent.com/42321264/199717599-165ac3e7-7f09-440b-b989-dc4761bff3cf.png)

If it also works as intended for you, then I will do some more aesthetic tweaks and I will add it to the next CRAN update!

A couple of notes though:
- You need to have R version higher or equal to 4.2.0.
- You need UCell installed.

Looking forward to your feedback!

Enrique

**Comment by enblacar** [comment]
Hi @cnk113,

Hmm, interesting. Does this happen, under the same parameters, when `flavor = "Seurat"`?  Can you maybe double-check if `group.by` is a metadata variable without NAs and `sigs` is a named vector?

I will have a look also to the code.

Best,
Enrique

**Comment by enblacar** [comment]
Hi @cnk113,

Glad to hear it worked for you! 

I do not fully understand what you are trying to suggest here, but it gave me another idea that might work as well as what you say. In the last commit, I implemented a `min.cutoff` and `max.cutoff` parameter, which will let you subset the color scale to set the min or max values to a given  one. The values have to be comprised in the absolute max or min values in the enrichment matrix.

This is a change affecting all the clusters at the same time, not a single one. But this is basically the aim of the visualization, to put the enrichment in a heatmap and compare between clusters.

Here you have some examples:

```r
# Original heatmap.
SCpubr::do_EnrichmentHeatmap(sample = sample, 
                             input_gene_list = genes, 
                             ncores = 1, 
                             flavor = "UCell", 
                             storeRanks = TRUE)
```
![78897747-9f19-42ea-ab2e-7f95690b52d0](https://user-images.githubusercontent.com/42321264/200115114-f34ac7d0-29e9-4249-bbdc-7ec8f5f134a5.png)

```r
# Set a max value for the scale.
SCpubr::do_EnrichmentHeatmap(sample = sample, 
                             input_gene_list = genes, 
                             ncores = 1, 
                             flavor = "UCell", 
                             storeRanks = TRUE,
                             max.cutoff = 0.5)
```
![9d7bbcdb-07d6-49c8-9cee-b2e5f5642abb](https://user-images.githubusercontent.com/42321264/200115136-ec444cf2-044f-481a-8acb-e026c7488411.png)

```r
# Set a min value for the scale.
SCpubr::do_EnrichmentHeatmap(sample = sample, 
                             input_gene_list = genes, 
                             ncores = 1, 
                             flavor = "UCell", 
                             storeRanks = TRUE,
                             min.cutoff = 0.5)
```
![194768a3-e3f7-4bbe-81fb-3364fa23a116](https://user-images.githubusercontent.com/42321264/200115157-371a736d-10c6-41f8-a6ca-7dfa155269e3.png)

```r
# Set a min and a max value for the scale
SCpubr::do_EnrichmentHeatmap(sample = sample, 
                             input_gene_list = genes, 
                             ncores = 1, 
                             flavor = "UCell", 
                             storeRanks = TRUE,
                             min.cutoff = 0.2,
                             max.cutoff = 0.5)
```
![d60cdeaf-117c-429b-aa77-69ea3c77fc0c](https://user-images.githubusercontent.com/42321264/200115187-536ccda2-a878-4fc6-a3b7-a8591fecc95e.png)

Please install the commit using:

```r
devtools::install_github("enblacar/SCpubr", ref = "52429ab0e0dc1d99450f6807bb840f392a577e08")
```

And be aware this is a dev version so the code might break. Let me know if this works for you! The changes will appear in the next CRAN update.

Best,
Enrique

**Comment by enblacar** [comment]
Hi @cnk113,

Happy to hear it worked for you! I have added `min.cutoff` and `max.cutoff` behaviour to all the rest of heatmap functions. 

You can access this by:

```r
devtools::install_github("enblacar/SCpubr", ref = "c6e4f7bbc635d49ec349c7b2c4832a5e144745e1")
```

All the changes will become part of the next CRAN update.

Closing this Issue for now. Thanks for your feedback!
Enrique

**Comment by enblacar** [comment]
All the changes have been implemented in SCpubr v1.0.3, currently available on CRAN.

Best,
Enrique
